src/sctools/gtf.py

```diff
@@ -0,0 +1,147 @@
+from collections import Iterable
+import string
+from . import reader
+
+
+class Record:
+    """
+    Simple namespace object that makes the fields of a GTF record available. Subclassed
+    to create records specific to exons, transcripts, and genes
+    """
+
+    __slots__ = ['_fields', '_attribute']
+
+    _del_letters = string.ascii_letters
+    _del_non_letters = ''.join(
+        set(string.printable).difference(string.ascii_letters))
+
+    def __init__(self, record):
+        """
+        :param str record: input record from file
+        """
+        fields = record.strip(';\n').split('\t')
+        self._fields = fields[:8]
+        self._attribute = {
+            key: value.strip('"') for (key, value) in
+            [field.split() for field in fields[8].split('; ')]
+        }
+
+    def __repr__(self):
+        return '<Record: %s>' % self.__str__()
+
+    def __bytes__(self):
+        return self.__str__().encode()
+
+    def __str__(self):
+        return '\t'.join(self._fields) + self._format_attribute() + '\n'
+
+    def __hash__(self) -> int:
+        return hash(self.__str__())
+
+    def _format_attribute(self):
+        return ' '.join('%s "%s";' % (k, v) for k, v in self._attribute.items())
+
+    @property
+    def seqname(self):
+        return self._fields[0]
+
+    @property
+    def chromosome(self):
+        return self._fields[0]  # synonym for seqname
+
+    @property
+    def source(self):
+        return self._fields[1]
+
+    @property
+    def feature(self):
+        return self._fields[2]
+
+    @property
+    def start(self):
+        return int(self._fields[3])
+
+    @property
+    def end(self):
+        return int(self._fields[4])
+
+    @property
+    def score(self):
+        return self._fields[5]
+
+    @property
+    def strand(self):
+        return self._fields[6]
+
+    @property
+    def frame(self):
+        return self._fields[7]
+
+    @property
+    def size(self):
+        size = self.end - self.start
+        if size < 0:
+            raise ValueError('invalid record: negative size %d (start > end)' % size)
```

Comment:
Why not calculate this in __init__ and throw an error there? That would protect against creation of invalid records. Or is there value in being able to parse invalid ones?

Reply:
Actually it's more about limiting unnecessary processing of records; I wanted to maximize speed here and so do no parsing of the records until the data is specifically requested. 

Often only a few fields of the record are of interest, or a particular type of record. For example, often times on gene records are wanted, and therefore there is no value in processing the much more frequent exon or transcript records. 

I grudgingly added the variable field parsing to `init` (`field[8:]`, which you correctly wanted an explanation for) because it contains gene id information that is normally of interest.

Reply:
Added explanation in docstring.